### PR TITLE
bazel: make lint pass unit tests runnable in the sandbox

### PR DESCRIPTION
**Pull request by rickystewart**
Some unit tests, these lint tests being an example, call into the `go`
binary directly. This is a problem in sandboxed environments where there
may be no globally installed `go` toolchain.

A summary of the changes made here:

1. We depend on the `rules_go` Go library, which surfaces some utilities
   that are useful here as well as generally for porting tests to work
   in Bazel.
2. Add a helper library in `pkg/build/bazel` to surface that stuff.
   There's a stub version of the library as well for code built without
   Bazel; we use build tags to make sure the right version of the
   library is surfaced, and update `.bazelrc` accordingly.
3. Refactor `testutils.TestDataPath` to use the helper lib -- this was
   almost entirely duplicated logic that's outmoded by `rules_go`.
4. Update all the tests to a) depend on the Go SDK when building, and
   b) use the helper library to patch in the toolchain.
5. Finally, delete a bunch of stray `BUILD.bazel` files in the
   `pkg/testutils/lint/passes` directory -- these serve no purpose and
   an old Gazelle configuration put them there. We just forgot to clean
   them up until now.

To test:
  dev test pkg/testutils/lint/passes/... -v

Release justification: Non-production code changes
Release note: None

Co-authored-by: irfan sharif <irfanmahmoudsharif@gmail.com>

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -12,6 +12,7 @@ load("@bazel_gazelle//:def.bzl", "gazelle")
 #
 # gazelle:prefix github.com/cockroachdb/cockroach
 # gazelle:build_file_name BUILD.bazel
+# gazelle:build_tags bazel
```
Comment (irfansharif) [inline BUILD.bazel:15]: Doesn't count for much but LGTM. If anything, maybe we could use a more greppable build tag here should things get unwieldy later on.

**Review by irfansharif — APPROVED** [review]

**Comment by rickystewart** [comment]
Still broken -- give me a bit to resolve.

**Comment by rickystewart** [comment]
OK, I think that should be everything.

**Comment by rickystewart** [comment]
Just needed to add `non_bazel.go` to a lint exclusion.

**Review by rail — APPROVED** [review]

**Comment by rickystewart** [comment]
TFTR!

bors r+

**Pull request merged**